internal/archive/archive.go

```diff
@@ -176,12 +183,40 @@ func openUbuntu(options *Options) (Archive, error) {
 	return archive, nil
 }
 
+func (index *ubuntuIndex) fetchVerifyReleaseFile() (io.ReadCloser, error) {
+	reader, err := index.fetch("InRelease", "", fetchDefault)
+	if err != nil {
+		return nil, err
+	}
+	content, err := io.ReadAll(reader)
+	reader.Close()
+	if err != nil {
+		return nil, err
+	}
+	block, _ := clearsign.Decode(content)
+	if block == nil {
+		return nil, fmt.Errorf("corrupted archive InRelease file: no clear-signed block")
+	}
+	err = pgperrors.ErrUnknownIssuer
+	for _, keyring := range index.archive.options.Keyrings {
+		_, err = block.VerifySignature(keyring, nil)
+		if err == nil || !errors.Is(err, pgperrors.ErrUnknownIssuer) {
+			break
+		}
+	}
+	if err != nil {
+		return nil, fmt.Errorf("release signature verification failed: %w", err)
+	}
+	return io.NopCloser(bytes.NewReader(block.Plaintext)), nil
+}
+
 func (index *ubuntuIndex) fetchRelease() error {
 	logf("Fetching %s %s %s suite details...", index.label, index.version, index.suite)
-	reader, err := index.fetch("Release", "", fetchDefault)
+	reader, err := index.fetchVerifyReleaseFile()
```

Comment:
Applied with one small variation that instead of passing around a reader, a byte array is passed, because `clearsign.Decode()` expects `[]byte`. If I passed around reader, I'd have to read it fully in `verifyRelease` and then create byte reader for `parseRelease`. This way, network IO is isolated in `fetchRelease`. What do you think? If you liked your suggestion better, I will apply it.